File: 01-presentation-blog-examples/src/main/java/com/nalepa/demo/example07/Example07_Monitored_Many_CPU_QUEUE.java
package com.nalepa.demo.example07;

import java.time.Duration;
import java.time.LocalTime;
import java.util.LinkedList;

public class Example07_Monitored_Many_CPU_QUEUE {
    public static void main(String[] args) {
        LinkedList<Runnable> queue = new LinkedList<>();

        Runnable monitoredRunnable1 = new MonitoredRunnable(() -> simulateCpuCode(1));
        Runnable monitoredRunnable2 = new MonitoredRunnable(() -> simulateCpuCode(2));
        Runnable monitoredRunnable3 = new MonitoredRunnable(() -> simulateCpuCode(3));

        queue.add(monitoredRunnable1);
        queue.add(monitoredRunnable2);
        queue.add(monitoredRunnable3);

        while (!queue.isEmpty()) {
            Runnable runnable = queue.poll();
            runnable.run();
        }
    }

    static long simulateCpuCode(int index) {
        log("Start CPU code for: " + index);
        long startTime = System.nanoTime();
        long iteration = 0L;
        while (Duration.ofNanos(System.nanoTime() - startTime).getSeconds() < 5) {
            iteration++;
            Math.sqrt((double) iteration);
        }
        log("End CPU code for: " + index);
        return iteration;
    }

    static void log(String message) {
        System.out.println(Thread.currentThread().getName() + " : " + LocalTime.now() + " : " + message);
    }

    static class MonitoredRunnable implements Runnable {
        private final Runnable delegate;
        private final long runnableInstanceCreatedAt;

        public MonitoredRunnable(Runnable delegate) {
            this.delegate = delegate;
            this.runnableInstanceCreatedAt = System.nanoTime();
            log("Created MonitoredRunnable instance");
        }

        @Override
        public void run() {
            System.out.println();
            log("Queue wait time took: " + Duration.ofNanos(System.nanoTime() - runnableInstanceCreatedAt).toSeconds() + " s");

            long startExecution = System.nanoTime();
            delegate.run();
            log("Task execution took: " + Duration.ofNanos(System.nanoTime() - startExecution).toSeconds() + " s");
        }
    }
}

File: presentation-blog-examples/src/main/kotlin/com/nalepa/demo/example07/Example07_Monitored_Many_CPU_QUEUE.kt
package com.nalepa.demo.example07

import java.time.Duration
import java.time.LocalTime
import java.util.*
import kotlin.math.sqrt

fun main() {
    val queue = LinkedList<Runnable>()

    // yup, we can write: queue.add { simulateBlockingIO(1) } but Runnable is intentional :P
    repeat(4) { index ->
        queue.add(MonitoredRunnable { simulateCpuCode(index) })
    }

    while (!queue.isEmpty()) {
        val runnable = queue.poll()
        runnable.run()
    }
}

fun simulateCpuCode(index: Int): Long {
    log("Start CPU code for: $index")
    val startTime = System.nanoTime()
    var iteration = 0L
    while (Duration.ofNanos(System.nanoTime() - startTime).seconds < 5) {
        // whatever cpu is doing, compression, JSON serialization, deserialization, etc
        iteration++
        sqrt(iteration.toDouble())
    }
    log("End CPU code for: $index")
    return iteration
}

fun log(message: String) {
    println("${Thread.currentThread().name} : ${LocalTime.now()} : $message")
}

class MonitoredRunnable(
    private val delegate: Runnable,
) : Runnable {

    val runnableInstanceCreatedAt = System.nanoTime()

    init {
        log("Created MonitoredRunnable instance")
    }

    override fun run() {
        println("")
        log("Queue wait time took: ${Duration.ofNanos(System.nanoTime() - runnableInstanceCreatedAt).toSeconds()} s")

        val startExecution = System.nanoTime()
        delegate.run()
        log("Task execution took: ${Duration.ofNanos(System.nanoTime() - startExecution).toSeconds()} s")
    }
}

